### ISSUE TITLE
Pass parameters to onSortEnd function

### ISSUE THREAD

**Issue by Adel-J** [clauderic/react-sortable-hoc#494]
I'm trying to pass a parameters to onSortEnd function but always the data is empty for mine but the indexing work fine.

`onSortEndCustom = ({oldIndex, newIndex}, data) => {
   console.log(data);
}`

I create my custom function it is working but seems call every time the list changes even if not change the sort, I want same concept as onSortEnd.

So any one know why it is not working to pass parameters or any best solutions?

**Comment by clauderic** [comment]
I'm not really sure I understand your question or the problem. Could you rephrase the question or share a CodeSandbox replicating the issue? Closing this until more info is provided

**Comment by WikeC** [comment]
I think I understand what he means because I meet the same case.
Here is my case:
I have a list group. How to change the order of the list inside the group?

Here is my example:
groupItems: [
{title: "Title 1", listItems: ["item 1.1", "item 1.2", "item 1.3"]}, {title: "Title 2", listItems: ["item 2.1", "item 2.2", " item 2.3 "]}
]

If I want to change the order of the list in the second array, (groupItems [1] .listItems) I need to tell the onSortEnd the id of the array to modify (for example):

onSortEnd = {() => this.onChangeItem (idTable)}

And in the function:

onChangeItem = ({oldIndex, newIndex}, idTable) => {
 console.log (idTable) // undefined
}

What would be the solution?

Thank you for your reply.

**Comment by WikeC** [comment]
> I think I understand what he means because I meet the same case.
> Here is my case:
> I have a list group. How to change the order of the list inside the group?
> 
> Here is my example:
> groupItems: [
> {title: "Title 1", listItems: ["item 1.1", "item 1.2", "item 1.3"]}, {title: "Title 2", listItems: ["item 2.1", "item 2.2", " item 2.3 "]}
> ]
> 
> If I want to change the order of the list in the second array, (groupItems [1] .listItems) I need to tell the onSortEnd the id of the array to modify (for example):
> 
> onSortEnd = {() => this.onChangeItem (idTable)}
> 
> And in the function:
> 
> onChangeItem = ({oldIndex, newIndex}, idTable) => {
> console.log (idTable) // undefined
> }
> 
> What would be the solution?
> 
> Thank you for your reply.

Ok got it ! 
Need to use the property collection.

**Comment by OscarLeung19920222** [comment]
@WikeC Hi, I have found this issue and it seem likes you have got the idea of how to solve it.
I am currently encountering same problem with the onSortEnd function.
Would you advise a bit more details on the solution? Really appreciated your help.

**Comment by WikeC** [comment]
@OscarLeung19920222 , Hi.
In your loop, you must assign your id to collection.

For example,

`      {items.map((value, index) => (
        <SortableItem key={`item-${index}`} index={index} value={value} collection={myId} />
      ))}`

And in your function,

`onSortEnd = ({oldIndex, newIndex, collection}) => {
  console.log(collection) // myId
}`


Hope I helped you.

**Comment by XeniaSiskaki** [comment]
I don't think the `collection` parameter is used for the use case that @Adel-J mentioned in his first comment. From the docs:

> The collection the element is part of. This is useful if you have multiple groups of sortable elements within the same SortableContainer

which if I understand correctly means that the value will just be an identifier for the group and not some custom data.

@clauderic I'm facing the same problem. I want to pass some custom data and read them in the `onSortEnd`. Is that currently possible? To be more specific, each `SortableElement` in my `SortableContainer` has an ID which I need to know when handling the sort end.

**Comment by sagar-gavhane** [comment]
Yes. You can pass a parameter to `onSortEnd` method using `collection` but it's only accepting number and string.

```jsx
<SortableList
  ...props
  collection={myId}
/>
```

```jsx
handleSortEvent = ({ oldIndex, newIndex, collection }, e) => {
   // here, collection=myId
  ...code
}
```

**Comment by sercanov** [comment]
@XeniaSiskaki is right, when you pass `collection` attribute to `SortableItem`, sorting brokes. I also need to get item ID ( or any other data ) in `onSortEnd` function. The only way seems to work is adding custom attribute to `SortableItem`, and try to parse it in `onSortEnd`'s `nodes` property. Like this;

```javascript
const SortableList = SortableContainer(
  ({ ids = [], data = [], onClick, onEdit, selected, ...props }) => {
    if (data)
      return (
        <List component="nav">
          {ids.map((id, index) => (
            <SortableItem
              key={id}
              index={index}
              itemID={id} // This custom field
              value={data[id]}
            />
          ))}
        </List>
      );
  }
);
```

And in `onSortEnd` trying to parse `itemID` ;

<img width="1061" alt="Screen Shot 2020-05-24 at 19 21 09" src="https://user-images.githubusercontent.com/10291644/82759149-24035280-9df4-11ea-93d4-97d13f8a5e33.png">

**Comment by VanKerkhovenSimon** [comment]
@sercanov Thank you for your advice but i can't find the custom attributes properties as you do.

My code look like this : 

````
import {sortableContainer,sortableElement} from 'react-sortable-hoc'

const SortableItem = sortableElement(({value}) => 
<li>{value.title}</li>
)

const SortableContainer = sortableContainer(({children}) => {
    return <ul>{children}</ul>
})

````

```
    onSortEnd = ({oldIndex,newIndex,collection,nodes, ...rest}) => {
       
        console.log(nodes[oldIndex].node.attributes)
}
```
```
 <SortableContainer onSortEnd={this.onSortEnd} >
              {this.state.possibilites.map((suggestion,index) => (
                    <SortableItem key={`item-${suggestion.id}`} index={index} item={suggestion} value={suggestion} />
              ))}
          </SortableContainer>
```
My suggestion data 
```
suggestion : {
id,
title,
more data...
}
```

Here is what i get from the console 
![Capture d’écran 2020-08-03 à 09 52 52](https://user-images.githubusercontent.com/56828845/89158966-2100ad00-d56f-11ea-9d8b-16e54e37ba2a.png)

What did i forget ?

**Comment by PCPbiscuit** [comment]
Found a much better solution to this. I needed the id of an item so I was also trying to pass it using 'collection' prop however it does indeed break the whole dnd. 
My solution to this is to use oldIndex to access the current item in the state:

`
onSortEnd ({oldIndex, newIndex}) {
        let id = this.state.items[oldIndex].id
`

**Comment by Anisha95** [comment]
As per my observation, if we pass params to the onSortEnd like this 

``` 
<SortableList
  ...props
  onSortEnd={(e) => onSortEnd({e: e, myId: myId}}
/>
```

Then on the function, the prop field myId could be received :
```
onSortEnd = (props, data) => {
     let myId = props.myId;  // props => {e: {collection: 0, newIndex: [newIndex], oldIndex: [oldIndex], isKeySorting: undefined, 
                                           //nodes: 
                                           // Array(2)}
                                          // myId: myId}
}

**Comment by chengwang1009** [comment]
I don't know am I too late for this. 

But here is the solution to pass data as props 

`const onSortEndHandler = ({oldIndex, newIndex, ...props}) => {

      const questionsArray = props.questionsArray;
      const setQuestionsArray = props.setQuestionsArray;

      const arr = arrayMoveImmutable(questionsArray, oldIndex, newIndex);
      setQuestionsArray(arr);
    };
    
    <SortableList 
            items={questionsArray}
	    onSortEnd={({oldIndex, newIndex})=>{onSortEndHandler({oldIndex, newIndex, questionsArray, setQuestionsArray})}}
            axis="xy"
     />`
